File: backend/src/main/resources/js/farm.js
function Farm(exportData) {
    var FIELD_HEIGHT = 10;
    var FIELD_WIDTH = 10;
    var farm;
    if (exportData && Array.isArray(exportData)) {
        farm = exportData;
    } else {
        farm = [];
        for (var i = 0; i < FIELD_HEIGHT; i++) {
            farm.push([]);
            for (var j = 0; j < FIELD_WIDTH; j++) {
                farm[i].push(staticData.getItems().ground);
            }
        }
    }

    this.applyHandToCell = function (hand, target) {
        if (!hand || !hand.id) {
            if (!hand) hand = {};
            throw "Invalid object in hand: " + JSON.stringify(hand);
        }
        if (!target || target.x === undefined || target.y === undefined) {
            if (!target) target = {};
            throw "Invalid target object: " + JSON.stringify(target);
        }
        if (bag.getOrCreate(hand.id).count <= 0) {
            throw "Not enough '" + hand.id + "'. Target: " + JSON.stringify(target) + ". Cell: " + JSON.stringify(farm[target.x][target.y]);
        }
        var item = staticData.getItems()[hand.id];
        farm[target.x][target.y] = item.use(farm[target.x][target.y]);
        bag.decreaseCount(hand.id, 1);
    };

    function updateTimersDeltas() {
        for (var x = 0; x < farm.length; x++) {
            for (var y = 0; y < farm[x].length; y++) {
                if (!farm[x][y]) {
                    continue;
                }
                var delta = farm[x][y].endTime - new Date().getTime();
                if (delta > 0) {
                    farm[x][y].currentProductionTimeLeft = delta;
                } else {
                    farm[x][y].currentProductionTimeLeft = 0;
                }
            }
        }
    }

    this.getOriginalFarmObject = function () {
        updateTimersDeltas();
        return farm;
    }
}

var farm = new Farm();

File: farmBackend/src/main/resources/lua/farm.lua
function Farm(exportData)
    local width = 5;
    local height = 7;
    local farm;

    if exportData and type(exportData) == "table" then
        farm = exportData
    else
        farm = {}
        for x = 1, height, 1 do
            farm[x] = {}
            for y = 1, width, 1 do
                farm[x][y] = staticData.getItems().ground
            end
        end
        farm[1][3] = staticData.getItems().carrotSpawnBox;
        farm[2][3] = staticData.getItems().road;
        farm[3][3] = staticData.getItems().road;
        farm[4][3] = staticData.getItems().road;
        farm[5][3] = staticData.getItems().road;
        farm[6][3] = staticData.getItems().road;
        farm[7][3] = staticData.getItems().road;
        farm[2][2] = staticData.getItems().field;
        farm[3][2] = staticData.getItems().field;
        farm[4][2] = staticData.getItems().field;
        farm[5][2] = staticData.getItems().field;
        farm[5][4] = staticData.getItems().well;
    end

    local applyHandToCell = function(hand, target, worker)
        if not hand or not hand.id then
            error("Invalid object in hand: " .. json.stringify(hand))
        end
        if not target or not target.x or not target.y then
            error("Invalid target object: " .. json.stringify(target))
        end
        local item = staticData.getItems()[hand.id]
        farm[target.x][target.y] = item.use(farm[target.x][target.y])
    end

    local updateTimerDeltas = function()
        for x = 1, #farm, 1 do
            for y = 1, #farm[x], 1 do
                if farm[x][y] and farm[x][y].endTime then
                    local delta = farm[x][y].endTime - os.time() * 1000
                    if delta > 0 then
                        farm[x][y].currentProductionTimeLeft = delta
                    else
                        farm[x][y].currentProductionTimeLeft = 0;
                    end
                end
            end
        end
    end

    local getOriginalFarmObject = function()
        updateTimerDeltas()
        return farm
    end

    return {
        width = width,
        height = height,
        applyHandToCell = applyHandToCell,
        getOriginalFarmObject = getOriginalFarmObject
    }
end